### PR TITLE
Add game modes.

### PR DESCRIPTION
**Pull request by patrycja93**
Introduced new abstract class and implementations for it due to the share code of normal games between two players and automatic tests. In the future will be posibility to play with a computer.

Documentation is not updated yet. Changed UIFactory class to return UI depending to the amount of input array parameters.

### CODE REVIEW

File: src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    @Override
+    public void settings(String[] inputArrayParameters) {
+        gameSettings = setup(inputArrayParameters);
+        Sequence sequence = new Sequence(gameSettings, inputArrayParameters[2]);
+        sequence.generateSequence();
+    }
+
+    private GameSettings setup(String[] inputArrayParameters) {
+        checkCorrectInputData(inputArrayParameters);
+        //TODO: generate draw sequence!!
```
Comment (BulaMar) [inline src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java]: Should this be here? Where are you generating other sequences?

File: src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    @Override
+    public void settings(String[] inputArrayParameters) {
+        gameSettings = setup(inputArrayParameters);
+        Sequence sequence = new Sequence(gameSettings, inputArrayParameters[2]);
+        sequence.generateSequence();
+    }
+
+    private GameSettings setup(String[] inputArrayParameters) {
+        checkCorrectInputData(inputArrayParameters);
```
Comment (BulaMar) [inline src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java]: Are you checking if input data is correct or checking correct input data?

File: src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        //TODO: generate draw sequence!!
+        return updateValuesInSettings(inputArrayParameters);
+    }
+
+    private GameSettings updateValuesInSettings(String[] inputArrayParameters) {
+        GameSettings.GameSettingsBuilder gameSettings = GameSettings.builder();
```
Comment (BulaMar) [inline src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java]: Could you collect all lines of builder into return statement and build it then?
Also this is not `gameSettings` but `gameSettingsBuilder` - not important when you will apply first line of this comment.

File: src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+        int boardSize = Integer.parseInt(inputArrayParameters[0]);
+        int unbrokenLine = Integer.parseInt(inputArrayParameters[1]);
+        int maxBoardSize = Math.max(boardSize, unbrokenLine);
+        int minUnbrokenLine = Math.min(boardSize, unbrokenLine);
+        gameSettings.boardSize(maxBoardSize);
+        gameSettings.unbrokenLine(minUnbrokenLine);
+
+        int towards = Integer.parseInt(inputArrayParameters[2]);
```
Comment (BulaMar) [inline src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java:55]: Please move this line under `int maxBoardSize = Math.max(boardSize, unbrokenLine);` that all `inputArrayParameters` values are in one place.

File: src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            case 0:
+            case 1: {
+                return numberOfMatchesHorizontalAndVertical(boardSize, unbrokenLine);
+            }
+            case 2:
+            case 3: {
+                return numberOfMatchesDiagonals(boardSize, unbrokenLine);
+            }
+            default:
+                return 3;
```
Comment (BulaMar) [inline src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java]: Please make constants with meaningful names from this cases.
For example: HORIZONTAL = 0

Additionally if you will already have it as constants try change switch case into map with functions.

File: src/main/java/pl/patrycja/ox/gameflow/FactoryMode.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    private int inputParametersSize;
+
+    FactoryMode(int inputParametersSize) {
+        this.inputParametersSize = inputParametersSize;
+    }
+
+    //create game settings object and set board size
+    //unbroken line, ui taking amount of args
+    Mode setMode() {
+        switch (inputParametersSize) {
```
Comment (BulaMar) [inline src/main/java/pl/patrycja/ox/gameflow/FactoryMode.java]: Try to replace with map. You do not want to adding each time case condition if new mode appear.

File: src/main/java/pl/patrycja/ox/gameflow/Game.java [hunk 7]
```diff
@@ -2,31 +2,45 @@
...
+        return updateValuesForBoardAndUnbrokenLine(inputArrayParameters);
+    }
 
-        for (int i = 0; i < gameSettings.getNumberOfMatches(); i++) {
-            List<Spectator> spectators = SpectatorsRoom.addSpectators(gameSettings);
-            Match.init(gameSettings, spectators)
-                    .addController(playerChanger)
-                    .start();
+    private GameSettings updateValuesForBoardAndUnbrokenLine(String[] inputArrayParameters) {
+        GameSettings.GameSettingsBuilder gameSettings = GameSettings.builder();
```
Comment (BulaMar) [inline src/main/java/pl/patrycja/ox/gameflow/Game.java]: Same situation with builder here.

File: src/main/java/pl/patrycja/ox/gameflow/Game.java [hunk 8]
```diff
@@ -2,31 +2,45 @@
...
+        int firstParameter = Integer.parseInt(inputArrayParameters[0]);
```
Comment (BulaMar) [inline src/main/java/pl/patrycja/ox/gameflow/Game.java]: Is that board size, maybe worth to name it like that?

File: src/main/java/pl/patrycja/ox/gameflow/Game.java [hunk 9]
```diff
@@ -2,31 +2,45 @@
...
+        if (inputArrayParameters.length == 1) {
+            gameSettings.boardSize(Math.max(firstParameter, 3));
+        } else {
+            int secondParameter = Integer.parseInt(inputArrayParameters[1]);
```
Comment (BulaMar) [inline src/main/java/pl/patrycja/ox/gameflow/Game.java]: unbroken line?

File: src/main/java/pl/patrycja/ox/gameflow/Mode.java [hunk 10]
```diff
@@ -2,77 +2,40 @@
...
-        String towards = args[2];
+    void play(List<Player> players) {
+        gameSettings.setPlayer();
+        PlayerChanger playerChanger = new PlayerChanger(players, gameSettings);
 
-        if (validateArguments(boardSize, unbrokenLine)) {
-            setBoardSizeAndUnbrokenLine(gameSettings, boardSize, unbrokenLine, towards);
+        for (int i = 0; i < gameSettings.getNumberOfMatches(); i++) {
+            List<Spectator> spectators = SpectatorsRoom.addSpectators(gameSettings);
+            //TODO: changes game setting on boardSize
```
Comment (BulaMar) [inline src/main/java/pl/patrycja/ox/gameflow/Mode.java]: What exactly are you going to do here?

File: src/main/java/pl/patrycja/ox/ui/InputChecker.java [hunk 11]
```diff
@@ -32,14 +28,14 @@ public int getValidNumber() {
         return Integer.parseInt(String.valueOf(input));
     }
 
-    public boolean getValidNumberDemoMode(String boardSize, String unbrokenLine) {
+    public void getValidNumberFromInputArray(String[] parameters) {
```
Comment (BulaMar) [inline src/main/java/pl/patrycja/ox/ui/InputChecker.java]: Will that still be a `get` type method?

File: src/main/java/pl/patrycja/ox/ui/InputChecker.java [hunk 12]
```diff
@@ -32,14 +28,14 @@ public int getValidNumber() {
...
-            Integer.parseInt(String.valueOf(boardSize));
-            Integer.parseInt(String.valueOf(unbrokenLine));
-            return true;
+            for (String p : parameters) {
+                Integer.parseInt(String.valueOf(p));
+            }
         } catch (IllegalArgumentException e) {
-            e.printStackTrace();
-            return false;
+            ui.display("Entered wrong arguments. Please enter integer numbers and run the game again.");
```
Comment (BulaMar) [inline src/main/java/pl/patrycja/ox/ui/InputChecker.java]: `Please run game again with correct integer numbers`?

File: src/main/java/pl/patrycja/ox/ui/UIFactory.java [hunk 13]
```diff
@@ -1,13 +1,32 @@
...
+    public UIFactory() {
+        this.kindsOfUi = setUi();
+    }
+
+    public UI getUiFor(int inputArraySize) {
+        UI ui = kindsOfUi.get(inputArraySize);
+        return ui;
+    }
+
+    private Map<Integer, UI> setUi() {
```
Comment (BulaMar) [inline src/main/java/pl/patrycja/ox/ui/UIFactory.java]: 👍  good that this is not a switch this time but do you need 3 times `new ConsoleUI()`

**Review by BulaMar — CHANGES_REQUESTED** [review]
Generally nice start for extendable modes in future but still need some correction in this state.

File: src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java (hunk 1 above)
Comment (patrycja93) [reply src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java]: Yes, good remark. I deleted the method generating sequence but I forgot to change the place for the todo.

File: src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java (hunk 2 above)
Comment (patrycja93) [reply src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java]: Ok, I will change this name.

File: src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java (hunk 3 above)
Comment (patrycja93) [reply src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java]: Ok, I will move it to the return statement.

File: src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java (hunk 4 above)
Comment (patrycja93) [reply src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java:55]: Ok, I will do it.

File: src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java (hunk 5 above)
Comment (patrycja93) [reply src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java]: Do you mean constants inside the method?

Comment (patrycja93) [reply src/main/java/pl/patrycja/ox/gameflow/AutomaticTests.java]: I made it inside the class as a private final.

File: src/main/java/pl/patrycja/ox/gameflow/FactoryMode.java (hunk 6 above)
Comment (patrycja93) [reply src/main/java/pl/patrycja/ox/gameflow/FactoryMode.java]: Ok, I will change it.

File: src/main/java/pl/patrycja/ox/gameflow/Game.java (hunk 7 above)
Comment (patrycja93) [reply src/main/java/pl/patrycja/ox/gameflow/Game.java]: ok  I will change it.

File: src/main/java/pl/patrycja/ox/gameflow/Game.java (hunk 8 above)
Comment (patrycja93) [reply src/main/java/pl/patrycja/ox/gameflow/Game.java]: Yes, you are right.

File: src/main/java/pl/patrycja/ox/gameflow/Game.java (hunk 9 above)
Comment (patrycja93) [reply src/main/java/pl/patrycja/ox/gameflow/Game.java]: I will change it.

File: src/main/java/pl/patrycja/ox/gameflow/Mode.java (hunk 10 above)
Comment (patrycja93) [reply src/main/java/pl/patrycja/ox/gameflow/Mode.java]: I will explain to you personally 😄

File: src/main/java/pl/patrycja/ox/ui/InputChecker.java (hunk 11 above)
Comment (patrycja93) [reply src/main/java/pl/patrycja/ox/ui/InputChecker.java]: I will change it.

File: src/main/java/pl/patrycja/ox/ui/InputChecker.java (hunk 12 above)
Comment (patrycja93) [reply src/main/java/pl/patrycja/ox/ui/InputChecker.java]: Ok.

File: src/main/java/pl/patrycja/ox/ui/UIFactory.java (hunk 13 above)
Comment (patrycja93) [reply src/main/java/pl/patrycja/ox/ui/UIFactory.java]: Probably I do not need it. I will think about it through.

**Review by BulaMar — CHANGES_REQUESTED** [review]
Thanks for making all requested changes.
Please fix failing builds.

**Review by BulaMar — APPROVED** [review]

**Pull request merged**